pipeline/metadata/add_metadata.py

```diff
@@ -0,0 +1,96 @@
+"""Class for adding ip metadata to rows"""
+
+from __future__ import absolute_import
+
+import datetime
+from typing import List, Tuple, Iterator, Iterable
+
+import apache_beam as beam
+
+from pipeline.metadata.beam_metadata import DateIpKey, IP_METADATA_PCOLLECTION_NAME, ROWS_PCOLLECION_NAME, make_date_ip_key, merge_metadata_with_rows
+from pipeline.metadata.schema import BigqueryRow, IpMetadataWithKeys
+from pipeline.metadata.ip_metadata_chooser import IpMetadataChooserFactory
+
+
+class MetadataAdder():
+  """This class exists to provide access to the instantiated IpMetadata inside a beam operation."""
+
+  def __init__(self,
+               metadata_chooser_factory: IpMetadataChooserFactory) -> None:
+    self.metadata_chooser_factory = metadata_chooser_factory
+
+  def add_metadata(
+      self, rows: beam.pvalue.PCollection[BigqueryRow]
+  ) -> beam.pvalue.PCollection[BigqueryRow]:
+    """Add ip metadata to a collection of roundtrip rows.
+
+    Args:
+      rows: beam.PCollection[BigqueryRow]
+
+    Returns:
+      PCollection[BigqueryRow]
+      The same rows as above with with additional metadata columns added.
+    """
+
+    # PCollection[Tuple[DateIpKey,BigqueryRow]]
+    rows_keyed_by_ip_and_date = (
+        rows | 'key by ips and dates' >>
+        beam.Map(lambda row: (make_date_ip_key(row), row)).with_output_types(
+            Tuple[DateIpKey, BigqueryRow]))
+
+    # PCollection[DateIpKey]
+    # pylint: disable=no-value-for-parameter
+    ips_and_dates = (
+        rows_keyed_by_ip_and_date | 'get ip and date keys per row' >>
+        beam.Keys().with_output_types(DateIpKey))
+
+    # PCollection[DateIpKey]
+    deduped_ips_and_dates = (
+        # pylint: disable=no-value-for-parameter
+        ips_and_dates | 'dedup' >> beam.Distinct().with_output_types(DateIpKey))
+
+    # PCollection[Tuple[date,List[ip]]]
+    grouped_ips_by_dates = (
+        deduped_ips_and_dates | 'group by date' >>
+        beam.GroupByKey().with_output_types(Tuple[str, Iterable[str]]))
+
+    # PCollection[Tuple[DateIpKey,IpMetadataWithKeys]]
+    ips_with_metadata = (
+        grouped_ips_by_dates | 'get ip metadata' >> beam.FlatMapTuple(
+            self._add_ip_metadata).with_output_types(Tuple[DateIpKey,
+                                                           IpMetadataWithKeys]))
+
+    # PCollection[Tuple[Tuple[date,ip],Dict[input_name_key,List[BigqueryRow|IpMetadataWithKeys]]]]
+    grouped_metadata_and_rows = (({
+        IP_METADATA_PCOLLECTION_NAME: ips_with_metadata,
+        ROWS_PCOLLECION_NAME: rows_keyed_by_ip_and_date
+    }) | 'group by keys' >> beam.CoGroupByKey())
+
+    # PCollection[BigqueryRow]
+    rows_with_metadata = (
+        grouped_metadata_and_rows |
+        'merge metadata with rows' >> beam.FlatMapTuple(
+            merge_metadata_with_rows).with_output_types(BigqueryRow))
+
+    return rows_with_metadata
+
+  def _add_ip_metadata(
```

Comment:
I'd call this `_annotate_ips`, so it's clear what the subject is.

Reply:
done